packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,513 @@
+  ///           setState(() {
+  ///             currentValue = newValue;
+  ///           });
+  ///         },
+  ///         groupValue: currentValue,
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  final ValueChanged<T> onValueChanged;
+
+  @override
+  _SegmentedControlState<T> createState() => _SegmentedControlState<T>();
+}
+
+class _SegmentedControlState<T> extends State<SegmentedControl<T>> {
+  int _selectedIndex;
+  int _pressedIndex;
+  Map<T, int> keysToIndices = <T, int>{};
+  @override
+  void initState() {
+    super.initState();
+    for (int index = 0; index < widget.children.keys.length; index += 1) {
+      keysToIndices[widget.children.keys.elementAt(index)] = index;
+    }
+    if (widget.children.keys.contains(widget.groupValue)) {
+      _selectedIndex = keysToIndices[widget.groupValue];
+    }
+  }
+
+  void _onTapDown(T currentKey) {
+    setState(() {
+      _pressedIndex = keysToIndices[currentKey];
+    });
+  }
+
+  void _onTapUp(TapUpDetails event) {
+    setState(() {
+      _pressedIndex = null;
+    });
+  }
+
+  void _onTapCancel() {
+    setState(() {
+      _pressedIndex = null;
+    });
+  }
+
+  void _onTap(T currentKey) {
+    setState(() {
+      if (_selectedIndex != keysToIndices[currentKey]) {
+        widget.onValueChanged(currentKey);
+      }
+    });
+  }
+
+  @override
+  void didUpdateWidget(SegmentedControl<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.children.keys.contains(widget.groupValue)) {
+      _selectedIndex = keysToIndices[widget.groupValue];
+    } else {
+      _selectedIndex = null;
+      keysToIndices.clear();
+      for (int index = 0; index < widget.children.keys.length; index += 1) {
+        keysToIndices[widget.children.keys.elementAt(index)] = index;
+      }
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> _gestureChildren = <Widget>[];
+    for (T currentKey in widget.children.keys) {
+      final TextStyle textStyle = DefaultTextStyle.of(context).style.copyWith(
+        color: (_selectedIndex == currentKey) ?
+        CupertinoColors.white : CupertinoColors.activeBlue,
+      );
+      final IconThemeData iconTheme = IconThemeData(
+        color: (_selectedIndex == currentKey) ?
+        CupertinoColors.white : CupertinoColors.activeBlue,
+      );
+
+      Widget child = widget.children[currentKey];
+      child = GestureDetector(
+        onTapDown: (TapDownDetails event) {
+          _onTapDown(currentKey);
+        },
+        onTapUp: _onTapUp,
+        onTapCancel: _onTapCancel,
+        onTap: () {
+          _onTap(currentKey);
+        },
+        child: new IconTheme(
+          data: iconTheme,
+          child: DefaultTextStyle(
+            style: textStyle,
+            overflow: TextOverflow.ellipsis,
+            child: Semantics(
+              inMutuallyExclusiveGroup: true,
+              selected: _selectedIndex != null ? _selectedIndex == currentKey : false,
```

Comment:
`_selectedIndex == currentKey` will never be true. Add a test that would have caught this. :-)